README.md
# django-ranged-fileresponse

Django do not include a HTTP Ranged Response. There is a [Django ticket](https://code.djangoproject.com/ticket/22479)
for this, however no indication that this feature will be implemented soon.

The [original suggested fix](https://github.com/satchamo/django/commit/2ce75c5c4bee2a858c0214d136bfcd351fcde11d)
applies the code to Django's static view. This is a packaged version of that fix,
but uses a modified FileResponse, instead of applying it to Django's static view.

Forked originally from a project that only handles local files
This fork also include Ranged Responses from open Google Storage Blobs

## Status

~Maintained

## Usage

### Requirements

 * django >= 3.2

### Installation

    pip install -e git://github.com/cluster311/django-ranged-fileresponse#egg=django-ranged-fileresponse==0.2.0

### Running

#### Local files

```python
    from ranged_fileresponse.local import RangedLocalFileResponse

    def some_proxy_view(request):
        filename = 'myfile.wav'
        response = RangedLocalFileResponse(request, open(filename, 'r'), content_type='audio/wav')
        response['Content-Disposition'] = 'attachment; filename="%s"' % filename
        return response
```

#### Google Storage files

```python
    from ranged_fileresponse.google import RangedGoogleBlobResponse

    def some_proxy_view(request):
        filename = 'myfile.wav'
        response = RangedGoogleBlobResponse(
            request,
            "https://storage.googleapis.com/parlarispa-app-files/audios/s02e38-adria-mercader.mp3",  # the media URL
            content_type='audio/wav'
        )
        response['Content-Disposition'] = 'attachment; filename="%s"' % filename
        return response
```

### Signals

```python
from django.dispatch import receiver
from ranged_fileresponse import ranged_file_response_signal

@receiver(ranged_file_response_signal)
def chunk_received(sender, uid, start, reloaded, finished, **kwargs):
    # do something with this data
    # save_stats(uid=uid, start=start, reloaded=reloaded, finished=finished)
```

## Contributing

See the [CONTRIBUTING.md](CONTRIBUTING.md) file on how to contribute to this project.

## Contributors

See the [CONTRIBUTORS.md](CONTRIBUTORS.md) file for a list of contributors to the project.

## Roadmap

### Changelog

The changelog can be found in the [CHANGELOG.md](CHANGELOG.md) file.

### In progress

 * Maintaining

## Get in touch with a developer

If you want to report an issue see the [CONTRIBUTING.md](CONTRIBUTING.md) file for more info.

We will be happy to answer your other questions at opensource@wearespindle.com

## License

django-ranged-fileresponse is made available under the MIT license. See the [LICENSE file](LICENSE) for more info.


CHANGELOG.md
# Changelog

## Version 0.2.0 2022-06-25

 * Fix iterator

## Version 0.1.9 2022-06-25

 * Add logger

## Version 0.1.8 2022-01-02

 * Allow Google Storage blob (with public URLs)
 * Remove Travis CI build

## Version 0.1.7 2021-03-14

 * Allow file-like objects
## Version 0.1.2

 * Implemented pull request #3, thanks @sih4sing5hong5

## Version 0.1.1

 * Updated readme

## Version 0.1

 * Intial release


ranged_fileresponse/google.py
import logging

from django.core.exceptions import BadRequest
from django.http.response import FileResponse

from ranged_fileresponse import RangedResponse
from ranged_fileresponse.google_storage_file import RangedGoogleStorageFileReader


logger = logging.getLogger(__name__)


class RangedGoogleBlobResponse(FileResponse, RangedResponse):
    """
    This is a modified FileResponse that returns `Content-Range` headers with
    the response, so browsers that request a media URL from Google,
    can stream the response properly.
    """

    def __init__(self, request, media_url,
                 block_size=None,  # allow to change
                 unique_id=None,  # to follow chunks outside via signal
                 *args, **kwargs):
        """
        Constructor also requires a request, which
        checks whether range headers should be added to the response.

        Args:
            request(WGSIRequest): The Django request object.
            file (File): A file-like object.
        """
        logger.info(f'init RangedGoogleBlobResponse {media_url}')
        self.unique_id = unique_id

        if 'HTTP_RANGE' in request.META:
            # we need the size to define exactly the ranges but we don't have them
            # until we read the first chunk
            start, stop = self.get_base_ranges(request.META['HTTP_RANGE'])
        else:
            start = 0
            stop = 0

        logger.info(f'start:stop {start}:{stop}')

        self.ranged_file = RangedGoogleStorageFileReader(
            media_url,
            block_size=block_size,
            start=start,
            stop=stop,
            unique_id=self.unique_id,
            ranged_response=self,  # connected to allow child to send signals
        )
        size = self.ranged_file.size
        super().__init__(self.ranged_file, *args, **kwargs)

        final_start = self.ranged_file.start
        final_stop = self.ranged_file.stop

        # fix ranges using the file size and add them to the response
        if 'HTTP_RANGE' in request.META:
            self.add_range_headers(final_start, final_stop, size)

        signal_kwargs = dict(
            start=start,
            stop=stop,
            uid=self.unique_id,
            reloaded=True,  # the user (or navigator) asks for another part
            finished=False,
            http_range=request.META.get('HTTP_RANGE', None)
        )
        self.send_signal(**signal_kwargs)

    def get_base_ranges(self, header):
        """
        Parses a range header into a list of two-tuples (start, stop) where
        `start` is the starting byte of the range (inclusive) and
        `stop` is the ending byte position of the range (exclusive).

        Args:
            header (str): The HTTP_RANGE request header.

        Returns:
            (start, stop) tuple or (None, None) if the value of the header is not syntatically valid.
        
        Specs https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Range
        Syntax
            Range: <unit>=<range-start>-
            Range: <unit>=<range-start>-<range-end>
            Range: <unit>=<range-start>-<range-end>, <range-start>-<range-end>
            Range: <unit>=<range-start>-<range-end>, <range-start>-<range-end>, <range-start>-<range-end>
            Range: <unit>=-<suffix-length>

        Directives
            <unit>
            The unit in which ranges are specified. This is usually bytes.

            <range-start>
            An integer in the given unit indicating the beginning of the request range.

            <range-end>
            An integer in the given unit indicating the end of the requested range. This value is optional and, if omitted, the end of the document is taken as the end of the range.

            <suffix-length>
            An integer in the given unit indicating the number of units at the end of the file to return.
        """

        if not header or '=' not in header:
            return None, None

        start = None
        stop = None
        units, range_ = header.split('=', 1)
        units = units.strip().lower()

        if units != 'bytes':
            raise BadRequest('Only bytes ranges are supported')

        # We don't handle multipart byteranges, just the first one
        val = range_.split(',')[0].strip()
        if '-' not in val:
            raise BadRequest(f'HTTP_RANGE header is bad defined (missing "-"): {header}')
        
        start, stop = val.split('-', 1)
        if start != '' and not start.isnumeric():
            raise BadRequest(f'HTTP_RANGE header is bad defined (start is not numeric): {header}')
        if stop != '' and not stop.isnumeric():
            raise BadRequest(f'HTTP_RANGE header is bad defined (stop is not numeric): {header}')

        stop = 0 if stop == '' else int(stop)
        if start == '' and stop > 0:  # corner case
            # start must be (FILE_SIZE - stop) but we don't know the file size
            # stop must be -> FILE_SIZE but we don't know the file size
            # Define start as negative as a flag
            start = -int(stop)
            stop = 0
        else:
            start = 0 if start == '' else int(start)

        return start, stop

    def add_range_headers(self, start, stop, size):
        """
        Adds several headers that are necessary for a streaming file
        response, in order for Safari to play audio files. Also
        sets the HTTP status_code to 206 (partial content).

        Args:
            range_header (str): Browser HTTP_RANGE request header.
        """
        self['Accept-Ranges'] = 'bytes'

        if start >= size:
            # Requested range not satisfiable.
            self.status_code = 416
            return

        if stop >= size:
            stop = size

        self['Content-Range'] = 'bytes %d-%d/%d' % (start, stop - 1, size)
        self['Content-Length'] = stop - start
        self.status_code = 206


ranged_fileresponse/google_storage_file.py
import io
import logging
import requests
from google.resumable_media.requests import ChunkedDownload


logger = logging.getLogger(__name__)
MAX_DOWN_SIZE = 0  # 31 * 1024 * 1024

class RangedGoogleStorageFileReader(object):
    """
    Wraps a google storage blob object with an iterator that runs over part (or all) of
    the file defined by start and stop. Blocks of block_size will be returned
    from the starting position, up to, but not including the stop point.
    """

    def __init__(self, media_url, start=0, stop=0, block_size=1024 * 1024, unique_id=None, ranged_response=None):
        """
        Args:
            media_url (str): URL to the resource
            start (int): Where to start reading the file.
            stop (Optional[int]:float): Where to end reading the file.
                Defaults to infinity.
            block_size (Optional[int]): The block_size to read with.
        """
        logger.info(f'Init RangedGoogleStorageFileReader {media_url} {start}:{stop}')
        self.f = io.BytesIO()
        self.media_url = media_url

        # -----------------------------------------------------------------------------------------
        if start < 0:
            # specifica case when we ask for the last N bytes but we still don't know thw file size
            self.download = ChunkedDownload(
                media_url=self.media_url,
                chunk_size=1024,  # just 1K to get the file size,
                stream=io.BytesIO(),
                start=0,
                # headers={},
            )
            chunk = self.download.consume_next_chunk(transport=requests.Session())
            logger.info(f'chunk {chunk} (start<0)')
            size = self.download.total_bytes
            start = size + start  # start is negative
        # -----------------------------------------------------------------------------------------

        self.download = ChunkedDownload(
            media_url=self.media_url,
            chunk_size=block_size,
            stream=self.f,
            start=start,
            # headers={},
        )
        logger.info(f'download {self.download} block {block_size} start {start}')

        # total_bytes is only available after the first chunk is downloaded
        self.initial_chunk = self.download.consume_next_chunk(transport=requests.Session())
        self.size = self.download.total_bytes
        if stop == 0:
            self.stop = self.size
        else:
            self.stop = stop

        self.block_size = block_size
        self.start = start
        self.position = self.start
        self.iter_counter = 0  # just to check

        # ensure a group of chunks with a max size (32 MB limit for http 1)
        if MAX_DOWN_SIZE > 0 and self.size - self.start > MAX_DOWN_SIZE:
            self.stop = self.start + MAX_DOWN_SIZE

        # only if client want something specific, use it, if not, go to the end
        if self.stop > 0:
            self.download.end = self.stop

        self.unique_id = unique_id

        # optionally, a father response to notify chenks sent
        self.ranged_response = ranged_response

    def __len__(self):
        return 1 + (self.size // self.block_size)

    def __iter__(self):
        return self

    def __next__(self):
        """
        Reads the data in chunks.
        """
        self.iter_counter += 1
        self._notify_chunk()
        logger.info(f'ITER {self.iter_counter} {self.position}-{self.block_size}:{self.stop}')
        self.position += self.block_size
        if self.iter_counter == 1:
            return self.initial_chunk.content
        
        if self.download.finished:
            logger.info(f'ITER down finished')
            raise StopIteration() 

        chunk = self.download.consume_next_chunk(transport=requests.Session())
        data = chunk.content
        
        if not data:
            logger.info(f'ITER data finished')
            raise StopIteration()

        return data
        
    def _notify_chunk(self):
        """ notify about a chunk """
        read_to = min(self.block_size, self.stop - self.position)
        my_stop = self.position + read_to
        
        if self.ranged_response:
            kwargs = dict(
                start=self.position,
                stop=my_stop,
                uid=self.unique_id,
                reloaded=False,
                finished=self.download.finished,
                http_range=None
            )
            self.ranged_response.send_signal(**kwargs)


ranged_fileresponse/local.py
from io import BufferedReader, BytesIO
import os
from django.http.response import FileResponse
from ranged_fileresponse import RangedResponse


class RangedFileReader(object):
    """
    Wraps a file like object with an iterator that runs over part (or all) of
    the file defined by start and stop. Blocks of block_size will be returned
    from the starting position, up to, but not including the stop point.
    """

    def __init__(self, file_like, start=0, stop=-1, block_size=1024 * 1024, unique_id=None, ranged_response=None):
        """
        Args:
            file_like (File): A file-like object.
            start (int): Where to start reading the file.
            stop (Optional[int]:float): Where to end reading the file.
                Defaults to infinity.
            block_size (Optional[int]): The block_size to read with.
        """
        self.f = file_like
        
        if type(self.f) == BufferedReader:  # a file from filesystem
            self.size = os.fstat(self.f.fileno()).st_size
        elif type(self.f) == BytesIO:  # a file like
            self.size = self.f.getbuffer().nbytes
        else:
            raise Exception('Unexpected buffer to stream ranged')

        self.block_size = block_size
        self.start = start
        self.unique_id = unique_id
        
        # the client frecuently do not ask for the stop bytes so we always send all the file
        # the client not uses all of this.
        # I try to force client to ask again
        if stop == -1:
            stop = min(start + stop, self.size)
        
        self.stop = stop
        self.ranged_response = ranged_response
        
    def __iter__(self):
        """
        Reads the data in chunks.
        """
        self.f.seek(self.start)
        position = self.start
        
        while position < self.stop:
            read_to = min(self.block_size, self.stop - position)
            data = self.f.read(read_to)
            my_stop = position + read_to
            if self.ranged_response:
                # notify about this chunk
                finished = position + self.block_size >= self.size  # size is real, stop is my idea to break in parts
                # we process this many times because the clients ask for start point and not the finsh ones
                # so we get chuck from 0 to end, then from N to end, then NN to end, etc.
                
                self.ranged_response.send_signal(
                    start=position,
                    stop=my_stop,
                    uid=self.unique_id,
                    reloaded=False,
                    finished=finished,
                    http_range=None
                )
            if not data:
                break
            yield data
            position += self.block_size

    def parse_range_header(self, header, resource_size):
        """
        Parses a range header into a list of two-tuples (start, stop) where
        `start` is the starting byte of the range (inclusive) and
        `stop` is the ending byte position of the range (exclusive).

        Args:
            header (str): The HTTP_RANGE request header.
            resource_size (int): The size of the file in bytes.

        Returns:
            None if the value of the header is not syntatically valid.
        """
        if not header or '=' not in header:
            return None

        ranges = []
        units, range_ = header.split('=', 1)
        units = units.strip().lower()

        if units != 'bytes':
            return None

        for val in range_.split(','):
            val = val.strip()
            if '-' not in val:
                return None

            if val.startswith('-'):
                # suffix-byte-range-spec: this form specifies the last N bytes
                # of an entity-body.
                start = resource_size + int(val)
                if start < 0:
                    start = 0
                stop = resource_size
            else:
                # byte-range-spec: first-byte-pos "-" [last-byte-pos].
                start, stop = val.split('-', 1)
                start = int(start)

                # The +1 is here since we want the stopping point to be
                # exclusive, whereas in the HTTP spec, the last-byte-pos
                # is inclusive.

                stop = int(stop) + 1 if stop else resource_size
                if start >= stop:
                    return None

            ranges.append((start, stop))

        return ranges


class RangedLocalFileResponse(FileResponse, RangedResponse):
    """
    This is a modified FileResponse that returns `Content-Range` headers with
    the response, so browsers that request the file, can stream the response
    properly.
    """

    def __init__(self, request, file,
                 block_size=None,  # allow to change
                 unique_id=None,  # to follow chunks outside via signal
                 max_content_size=0,  # not send all file, send parts so the client need to ask for more and I can do analytics
                 *args, **kwargs):
        """
        Constructor also requires a request, which
        checks whether range headers should be added to the response.

        Args:
            request(WGSIRequest): The Django request object.
            file (File): A file-like object.
        """
        self.unique_id = unique_id
        if max_content_size == 0:
            max_content_size = float('inf')
        
        self.ranged_file = RangedFileReader(file,
                                            block_size=block_size,
                                            stop=max_content_size,
                                            unique_id=self.unique_id)
        super().__init__(self.ranged_file, *args, **kwargs)
        if hasattr(file, 'close') and hasattr(self, '_closable_objects'):
            self._closable_objects.append(file)

        # default if we dont have range readed
        self.last_start = -1
        self.last_stop = -1
        
        if 'HTTP_RANGE' in request.META:
            self.add_range_headers(request.META['HTTP_RANGE'])

        start = self.last_start
        stop = self.last_stop
        # notify about this chunk
        self.send_signal(
            start=start,
            stop=stop,
            uid=self.unique_id,
            reloaded=True,  # the user (or navigator) asks for another part
            finished=False,
            http_range=request.META.get('HTTP_RANGE', None)
        )

    def add_range_headers(self, range_header):
        """
        Adds several headers that are necessary for a streaming file
        response, in order for Safari to play audio files. Also
        sets the HTTP status_code to 206 (partial content).

        Args:
            range_header (str): Browser HTTP_RANGE request header.
        """
        self['Accept-Ranges'] = 'bytes'
        size = self.ranged_file.size

        try:
            ranges = self.ranged_file.parse_range_header(range_header, size)
        except ValueError:
            ranges = None
        # Only handle syntactically valid headers, that are simple (no
        # multipart byteranges).
        if ranges is not None and len(ranges) == 1:
            start, stop = ranges[0]
            self.last_start = start
            self.last_stop = stop
            if start >= size:
                # Requested range not satisfiable.
                self.status_code = 416
                return

            if stop >= size:
                stop = size
            self.last_stop = stop

            self.ranged_file.start = start
            self.ranged_file.stop = stop
            self['Content-Range'] = 'bytes %d-%d/%d' % (start, stop - 1, size)
            self['Content-Length'] = stop - start
            self.status_code = 206
